src/txn/txn_rollback_to_stable.c

```diff
@@ -110,20 +45,194 @@ __txn_abort_newer_update(
+
+    WT_ERR(__wt_row_leaf_key(session, page, rip, key, false));
+
+    /* Open a history store table cursor. */
+    __wt_hs_cursor(session, &hs_cursor, &session_flags);
+    ret = __wt_hs_cursor_position(session, hs_cursor, hs_btree_id, key, WT_TS_MAX);
+    for (; ret == 0; ret = hs_cursor->prev(hs_cursor)) {
+        WT_ERR(hs_cursor->get_key(hs_cursor, &hs_btree_id, hs_key, &hs_start.timestamp,
+          &hs_start.txnid, &hs_stop.timestamp, &hs_stop.txnid));
+
+        /* Stop before crossing over to the next btree */
+        if (hs_btree_id != S2BT(session)->id)
+            break;
+
+        /*
+         * Keys are sorted in an order, skip the ones before the desired key, and bail out if we
+         * have crossed over the desired key and not found the record we are looking for.
+         */
+        WT_ERR(__wt_compare(session, NULL, hs_key, key, &cmp));
+        if (cmp != 0)
+            break;
+
+        /* Stop processing when we find the update that is less than the given the update */
+        if (hs_start.timestamp <= rollback_timestamp) {
+            valid_update_found = true;
+            break;
+        }
+
+        __wt_hs_store_time_pair(session, WT_TS_NONE, WT_TXN_NONE);
+        WT_ERR(hs_cursor->remove(hs_cursor));
+        WT_STAT_CONN_INCR(session, txn_rollback_hs_removed);
+    }
+
+    /*
+     * Found any history value that satisfy the given the timestamp, add it to the update list
+     * otherwise remove the key.
+     */
+    if (valid_update_found) {
+        WT_ERR(hs_cursor->get_value(hs_cursor, &durable_ts, &prepare_state, &type, hs_value));
+        WT_ERR(__wt_update_alloc(session, hs_value, &upd, &size, WT_UPDATE_STANDARD));
+        upd->txnid = hs_start.txnid;
+        upd->durable_ts = durable_ts;
+        upd->start_ts = hs_start.timestamp;
+    } else {
+        WT_ERR(__wt_update_alloc(session, NULL, &upd, &size, WT_UPDATE_TOMBSTONE));
+        upd->txnid = WT_TXN_NONE;
+        upd->durable_ts = WT_TS_NONE;
+        upd->start_ts = WT_TS_NONE;
+    }
+
+    /* If we don't yet have a modify structure, we'll need one. */
+    WT_ERR(__wt_page_modify_init(session, page));
+    mod = page->modify;
+
+    /* Allocate an update array as necessary. */
+    WT_PAGE_ALLOC_AND_SWAP(session, page, mod->mod_row_update, upd_entry, page->entries);
+
+    /* Set the WT_UPDATE array reference. */
+    upd_entry = &mod->mod_row_update[WT_ROW_SLOT(page, rip)];
+    upd->next = NULL;
+    WT_ERR(__wt_update_serial(session, page, upd_entry, &upd, size, false));
+
+    /* Finally remove that update from history store. */
+    if (valid_update_found) {
+        __wt_hs_store_time_pair(session, WT_TS_NONE, WT_TXN_NONE);
+        WT_ERR(hs_cursor->remove(hs_cursor));
+        WT_STAT_CONN_INCR(session, txn_rollback_hs_removed);
+    }
+
+err:
+    __wt_scr_free(session, &key);
+    __wt_scr_free(session, &hs_key);
+    __wt_scr_free(session, &hs_value);
+    if (ret != 0)
+        __wt_free(session, upd);
```

Comment:
I think it's more obvious to do something like this:
```
*** ../txn/txn_rollback_to_stable.c     Thu Jan 30 22:11:16 2020
--- -   Thu Jan 30 23:47:11 2020
*************** 
*** 166,171 ****
--- 166,172 ----
      upd_entry = &mod->mod_row_update[WT_ROW_SLOT(page, rip)];
      upd->next = NULL;
      WT_ERR(__wt_update_serial(session, page, upd_entry, &upd, size, false));
+     upd = NULL;

      /* Finally remove that update from history store. */
      if (valid_update_found) {
***************
*** 178,184 ****
      __wt_scr_free(session, &key);
      __wt_scr_free(session, &hs_key);
      __wt_scr_free(session, &hs_value);
!     if (ret != 0)
          __wt_free(session, upd);
      WT_TRET(__wt_hs_cursor_close(session, &hs_cursor, session_flags));
      
--- 179,185 ----
      __wt_scr_free(session, &key);
      __wt_scr_free(session, &hs_key);
      __wt_scr_free(session, &hs_value);
!     if (upd != NULL)
          __wt_free(session, upd);
      WT_TRET(__wt_hs_cursor_close(session, &hs_cursor, session_flags));
```

Reply:
Changed as you suggested.